tool/tsh/common/tlsmuxlistener.go

```diff
@@ -0,0 +1,93 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package common
+
+import (
+	"bufio"
+	"crypto/tls"
+	"net"
+
+	"github.com/gravitational/trace"
+)
+
+// connReader is a net.Conn wrapper with additional Peek() method.
+type connReader struct {
+	reader *bufio.Reader
+	net.Conn
+}
+
+// newBufferedConn is a connReader constructor.
+func newBufferedConn(conn net.Conn) connReader {
+	return connReader{
+		reader: bufio.NewReader(conn),
+		Conn:   conn,
+	}
+}
+
+// Peek reads n bytes without advancing the reader.
+// It's basically a wrapper around (bufio.Reader).Peek()
+func (b connReader) Peek(n int) ([]byte, error) {
+	return b.reader.Peek(n)
+}
+
+// Read returns data from underlying buffer.
+func (b connReader) Read(p []byte) (int, error) {
+	return b.reader.Read(p)
+}
+
+type tlsMuxListener struct {
+	listener  net.Listener
+	tlsConfig *tls.Config
+}
+
+func (m *tlsMuxListener) Accept() (net.Conn, error) {
+	conn, err := m.listener.Accept()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	bufConn := newBufferedConn(conn)
+	buf, err := bufConn.Peek(1)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	switch buf[0] {
+	case 0x16, 0x15:
```

Comment:
why `0x15`?

Reply:
Seen that with the older protocol versions, I think TLS 1.0?

Reply:
hmm. so oracle client that old 😆 ?
our main multiplexer only does `0x16`:
https://github.com/gravitational/teleport/blob/ec870e156a44e85998cee8049bdda167046a6196/lib/multiplexer/multiplexer.go#L742

up to you if you want to leave `0x15` here but lets add comment if so.

Reply:
I think `0x16` is enough on its own, I'll drop `0x15`.